daemon/libnetwork/portmappers/nat/mapper_linux.go

```diff
@@ -0,0 +1,325 @@
+package nat
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"net/netip"
+	"os"
+	"strconv"
+	"syscall"
+
+	"github.com/containerd/log"
+	"github.com/docker/docker/daemon/libnetwork/internal/rlkclient"
+	"github.com/docker/docker/daemon/libnetwork/portallocator"
+	"github.com/docker/docker/daemon/libnetwork/portmapperapi"
+	"github.com/docker/docker/daemon/libnetwork/types"
+)
+
+const (
+	driverName              = "nat"
+	maxAllocatePortAttempts = 10
+)
+
+type PortDriverClient interface {
+	ChildHostIP(hostIP netip.Addr) netip.Addr
+	AddPort(ctx context.Context, proto string, hostIP, childIP netip.Addr, hostPort int) (func() error, error)
+}
+
+type proxyStarter func(types.PortBinding, *os.File) (func() error, error)
+
+// Register the "nat" port-mapper with libnetwork.
+func Register(r portmapperapi.Registerer, cfg Config) error {
+	return r.Register(driverName, NewPortMapper(cfg))
+}
+
+type PortMapper struct {
+	// pdc is used to interact with rootlesskit port driver.
+	pdc         PortDriverClient
+	startProxy  proxyStarter
+	enableProxy bool
+}
+
+type Config struct {
+	// RlkClient is called by MapPorts to determine the ChildHostIP and ask
+	// rootlesskit to map ports in its netns.
+	RlkClient   PortDriverClient
+	StartProxy  proxyStarter
+	EnableProxy bool
+}
+
+func NewPortMapper(cfg Config) portmapperapi.PortMapper {
```

Comment:
Nit: return a concrete-typed value.

```suggestion
func NewPortMapper(cfg Config) PortMapper {
```